pkg/sarif.go

```diff
@@ -0,0 +1,446 @@
+	// Not supported by GitHub, but possibly useful.
+	PartialFingerprints partialFingerprints `json:"partialFingerprints,omitempty"`
+}
+
+type automationDetails struct {
+	ID string `json:"id"`
+}
+
+type run struct {
+	AutomationDetails automationDetails `json:"automationDetails"`
+	Tool              tool              `json:"tool"`
+	// For generated files during analysis. We leave this blank.
+	// See https://github.com/microsoft/sarif-tutorials/blob/main/docs/1-Introduction.md#simple-example.
+	Artifacts string   `json:"artifacts,omitempty"`
+	Results   []result `json:"results"`
+}
+
+type sarif210 struct {
+	Schema  string `json:"$schema"`
+	Version string `json:"version"`
+	Runs    []run  `json:"runs"`
+}
+
+func detailToRegion(details *checker.CheckDetail) region {
+	var reg region
+	var snippet *text
+	if details.Msg.Snippet != "" {
+		snippet = &text{Text: details.Msg.Snippet}
+	}
+	switch details.Msg.Type {
+	default:
+		panic("invalid")
+	case checker.FileTypeURL:
+		reg = region{
+			StartLine: &details.Msg.Offset,
+			Snippet:   snippet,
+		}
+	case checker.FileTypeNone:
+		// Do nothing.
+	case checker.FileTypeSource:
+		reg = region{
+			StartLine: &details.Msg.Offset,
+			Snippet:   snippet,
+		}
+	case checker.FileTypeText:
+		reg = region{
+			CharOffset: &details.Msg.Offset,
+			Snippet:    snippet,
+		}
+	case checker.FileTypeBinary:
+		reg = region{
+			ByteOffset: &details.Msg.Offset,
+		}
+	}
+	return reg
+}
+
+func shouldAddLocation(detail *checker.CheckDetail, showDetails bool,
+	logLevel zapcore.Level, minScore, score int) bool {
+	switch {
+	default:
+		return false
+	case detail.Msg.Path == "",
+		!showDetails,
+		logLevel != zapcore.DebugLevel && detail.Type == checker.DetailDebug,
+		detail.Msg.Type == checker.FileTypeURL:
+		return false
+	case score == checker.InconclusiveResultScore:
+		return true
+	case minScore < score && detail.Type == checker.DetailInfo:
+		return false
+	case minScore >= score:
+		return true
+	}
+}
+
+func shouldAddRelatedLocation(detail *checker.CheckDetail, showDetails bool,
+	logLevel zapcore.Level, minScore, score int) bool {
+	switch {
+	default:
+		return false
+	case detail.Msg.Path == "",
+		!showDetails,
+		detail.Msg.Type != checker.FileTypeURL,
+		logLevel != zapcore.DebugLevel && detail.Type == checker.DetailDebug:
+		return false
+	case score == checker.InconclusiveResultScore:
+		return true
+	case minScore < score && detail.Type == checker.DetailInfo:
+		return false
+	case minScore >= score:
+		return true
+	}
+}
+
+func detailsToLocations(details []checker.CheckDetail,
+	showDetails bool, logLevel zapcore.Level, minScore, score int) []location {
+	locs := []location{}
+
+	//nolint
+	// Populate the locations.
+	// Note https://docs.github.com/en/code-security/secure-coding/integrating-with-code-scanning/sarif-support-for-code-scanning#result-object
+	// "Only the first value of this array is used. All other values are ignored."
+	if showDetails {
```

Comment:
nit: invert and early return. then we can avoid the indentation of lines 227 -244

Reply:
good catch. I split this function in two in the last CL, but now you're right it can be simplified!